### PR TITLE
[DCOS-58676] Build automation tooling

### PR DESCRIPTION
**Pull request by alembiewski**
Resolves [DCOS-58676: Build automation tooling](https://jira.mesosphere.com/browse/DCOS-58676)

Changes:
- Add Dockerfile for builder image with required tools;
- Add/extend `make` targets;
- Add tag to `spark-on-k8s-operator` image;
- Refactoring

### CODE REVIEW

File: images/builder/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM golang:1.13.0@sha256:de697ce5ae02f3d9a57b0603fbb648efadfa212727e702ad3a807b43eba7f6d6
+
+RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.16.0/bin/linux/amd64/kubectl && \
```
[the remark this answers is not in the document]
Comment (alembiewski) [reply images/builder/Dockerfile]: Actually, `kubectl` is mandatory as well as `go` environment setup

File: Makefile [hunk 2]
```diff
@@ -37,28 +49,60 @@ cluster-destroy:
...
+		-v $(ROOT_DIR)/admin.conf:/root/.kube/config \
+		-e SPARK_IMAGE=$(SPARK_IMAGE_FULL_NAME) \
+		-e SPARK_OPERATOR_IMAGE=$(SPARK_ON_K8S_OPERATOR_IMAGE_FULL_NAME) \
+		$(shell cat docker-builder) \
+		/bin/bash -c \
+		"kubectl config use-context kubernetes-admin@kudo-spark-operator && \
+		kubectl cluster-info && \
+		echo \$$SPARK_IMAGE && \
+		echo \$$SPARK_OPERATOR_IMAGE && \
+		/run-tests.sh"
```
Comment (akirillov) [inline Makefile]: will that work well with CI in terms of all the environment variables passed to  `/run-tests.sh`?  We should either invoke `make test` from `/run-tests.sh` and do all the configuration here, or make sure CI will set all the same parameters needed for the script. The contract is that CI triggers only `/run-tests.sh`

File: Makefile [hunk 3]
```diff
@@ -37,28 +49,60 @@ cluster-destroy:
...
+
+.PHONY: docker-builder
+docker-builder:
+	if [[ -z "$(shell docker images -q $(DOCKER_BUILDER_IMAGE_FULL_NAME))" ]]; then
+		docker build \
+			-t $(DOCKER_BUILDER_IMAGE_FULL_NAME) \
+			-f ${DOCKER_BUILDER_DOCKERFILE_PATH}/Dockerfile \
+			${DOCKER_BUILDER_DOCKERFILE_PATH}
+	fi
+	echo $(DOCKER_BUILDER_IMAGE_FULL_NAME) > $@
```
Comment (akirillov) [inline Makefile]: it looks like this line can be moved inside the conditional above

File: Makefile [hunk 4]
```diff
@@ -37,28 +49,60 @@ cluster-destroy:
...
+	echo "${SPARK_IMAGE_FULL_NAME}" > $@
 
 .PHONY: docker-build
 docker-build: spark-build
 docker-build:
-	docker build \
-		--build-arg SPARK_IMAGE=$(shell cat spark-build) \
-		-t ${SPARK_ON_K8S_OPERATOR_IMAGE_NAME} \
-		-f ${SPARK_ON_K8S_OPERATOR_DOCKERFILE_PATH} ${SPARK_OPERATOR_DIR}
+	if [[ -z "$(shell docker images -q $(SPARK_ON_K8S_OPERATOR_IMAGE_FULL_NAME))" ]]; then
```
Comment (akirillov) [inline Makefile]: it's a neat move but don't we want to rely on `make` here and just `echo` the image name to the target file? Also, it doesn't look like Docker-related targets should be `PHONY`, we can introduce `clean-docker` target for the rebuild purposes.

**Comment by alembiewski** [comment]
Thanks for the review @akirillov! All the comments have been addressed in accordance with CI contract as well as local development workfolw, thanks for the input!

**Comment by alembiewski** [comment]
Thanks @akirillov ! It looks like this one could be merged to master. @rpalaznik, could you rebase on master and use docker builder to run tests in scope of #6?

**Pull request merged**